[JIT] Abstract tracer::enter and tracer::exit into a function

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28474 [JIT] Trace module calls
* **#28473 [JIT] Abstract tracer::enter and tracer::exit into a function**

Differential Revision: [D18121007](https://our.internmc.facebook.com/intern/diff/D18121007)

Commit:
[JIT] Abstract tracer::enter and tracer::exit into a function

[ghstack-poisoned]

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

zdevito commented on torch/csrc/jit/python_tracer.h:
```diff
@@ -27,7 +27,7 @@ Node* preRecordPythonTrace(
     std::vector<THPObjectPtr> scalar_args);
 
-std::shared_ptr<Graph> createGraphByTracing(
+std::pair<std::shared_ptr<TracingState>, Stack> createGraphByTracing(
```
Changing the return type here makes the API less clear. In many places the only thing used as an output of this is the graph. TracingState is an implementation detail, it should not leak into the python code at all.

zdevito replied:
I am pretty sure TracingState is never used when returned. Only the graph is used. It is still fine to return a pair of Graph, outputs for the few cases where we use the outputs after tracing. But let's stop exposing TracingState and delete the python bindings for it.

zdevito replied:
There are a bunch of places in onnx code where it appears that the trace is returned, but it always ultimately ends up returning the graph. So this can be refactored.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -597,8 +597,11 @@ void initJitScriptBindings(PyObject* module) {
             auto typed_inputs = toTraceableStack(input_tuple);
-            auto graph = tracer::createGraphByTracing(
+            std::shared_ptr<tracer::TracingState> state;
+            Stack outs;
```
outs, state are unused only graph is used

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -793,8 +796,11 @@ void initJitScriptBindings(PyObject* module) {
          bool force_outplace) {
         auto typed_inputs = toTraceableStack(input_tuple);
-        auto graph = tracer::createGraphByTracing(
+        std::shared_ptr<tracer::TracingState> state;
```
same here

zdevito commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -343,6 +345,39 @@ void exit(const Stack& outputs) {
+  try {
+    std::shared_ptr<TracingState> state;
+    Stack trace_state_inputs;
+    std::tie(state, trace_state_inputs) = tracer::enter(std::move(inputs), self);
```
tracer enter/exit should now be private. They should be marked static. I think they are small enough, that they can just be manually inlined here.

zdevito commented on torch/jit/__init__.py:
```diff
@@ -338,39 +338,37 @@ def forward(self, *args):
-            # the inputs before (inputs_states[0]) and after (inputs_states[1])
-            # the trace to verify if they were modified during the trace
+
+        state = {}
```
why not just 3 variables?

zdevito commented on torch/jit/__init__.py:
```diff
@@ -338,39 +338,37 @@ def forward(self, *args):
+
         if self._return_inputs:
-            return trace, out, ret_inputs
+            return trace, state['outs'], state['ret_inputs']
```
I bet that we do not need to actually expose the tracing state object here.

zdevito reviewed (commented):
Thanks for the cleanup. There is one extra bit here so that we do not make the createGraphByTracing API worse.

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"




[ghstack-poisoned]

zdevito reviewed (approved):
Looks good now!

Commit:
Update on "[JIT] Abstract tracer::enter and tracer::exit into a function"


Differential Revision: [D18121007](https://our.internmc.facebook.com/intern/diff/D18121007)

[ghstack-poisoned]
